src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,119 +287,198 @@ class HandshakeALotClosure : public HandshakeClosure {
-            (os::elapsedTime() > (double)SelfDestructTimer * 60.0)) {
-          tty->print_cr("VM self-destructed");
-          exit(-1);
-        }
-
-        // If the queue contains a safepoint VM op,
-        // clean up will be done so we can skip this part.
-        if (!_vm_queue->peek_at_safepoint_priority()) {
-
-          // Have to unlock VMOperationQueue_lock just in case no_op_safepoint()
-          // has to do a handshake when HandshakeALot is enabled.
-          MutexUnlocker mul(VMOperationQueue_lock, Mutex::_no_safepoint_check_flag);
-          if ((_cur_vm_operation = VMThread::no_op_safepoint()) != NULL) {
-            // Force a safepoint since we have not had one for at least
-            // 'GuaranteedSafepointInterval' milliseconds and we need to clean
-            // something. This will run all the clean-up processing that needs
-            // to be done at a safepoint.
-            SafepointSynchronize::begin();
-            SafepointSynchronize::end();
-            _cur_vm_operation = NULL;
-          }
-        }
-        _cur_vm_operation = _vm_queue->remove_next();
+void VMThread::wait_until_executed(VM_Operation* op) {
+  MonitorLocker ml(VMOperation_lock,
+                   Thread::current()->is_Java_thread() ?
+                     Mutex::_safepoint_check_flag :
+                     Mutex::_no_safepoint_check_flag);
+  {
+    TraceTime timer("Installing VM operation", TRACETIME_LOG(Trace, vmthread));
+    while (true) {
+      if (VMThread::vm_thread()->set_next_operation(op)) {
+        ml.notify_all();
+        break;
       }
+      // Wait to install this operation as the next operation in the VM Thread
+      log_trace(vmthread)("A VM operation already set, waiting");
+      ml.wait();
+    }
+  }
+  {
+    // Wait until the operation has been processed
+    TraceTime timer("Waiting for VM operation to be completed", TRACETIME_LOG(Trace, vmthread));
+    // _next_vm_operation is cleared holding VMOperation_lock after it have been
+    // executed. We wait until _next_vm_operation not our op.
+    while (_next_vm_operation == op) {
+      // VM Thread can process it once we unlocks the mutex on wait.
+      ml.wait();
+    }
+  }
+}
 
-      if (should_terminate()) break;
-    } // Release mu_queue_lock
+static void self_destruct_if_needed() {
+  // Support for self destruction
+  if ((SelfDestructTimer != 0) && !VMError::is_error_reported() &&
+      (os::elapsedTime() > (double)SelfDestructTimer * 60.0)) {
+    tty->print_cr("VM self-destructed");
+    exit(-1);
+  }
+}
 
-    //
-    // Execute VM operation
-    //
-    { HandleMark hm(VMThread::vm_thread());
+void VMThread::inner_execute(VM_Operation* op) {
+  assert(Thread::current()->is_VM_thread(), "Must be the VM thread");
+
+  VM_Operation* prev_vm_operation = NULL;
+  if (_cur_vm_operation != NULL) {
+    // Check the VM operation allows nested VM operation.
+    // This normally not the case, e.g., the compiler
```

Comment:
typo - s/Check the VM/Check that the VM/
typo - s/This normally not/This is normally not/

Reply:
Fixed